userdetails-cognito/src/main/groovy/au/org/ala/userdetails/CognitoUserService.groovy

```diff
@@ -858,4 +867,78 @@ class CognitoUserService implements IUserService<UserRecord, UserPropertyRecord,
         resultStreamer.complete()
     }
 
+    @Override
+    Map generateApikey(String usagePlanId) {
+        if(!usagePlanId){
+            return [apikeys:null, err: "No usage plan id to generate api key"]
+        }
+
+        CreateApiKeyRequest request = new CreateApiKeyRequest()
+        request.enabled = true
+        request.customerId = currentUser.userId
+        request.name = "API key for user " + currentUser.userId
+        def response = apiGatewayIdp.createApiKey(request)
+
+        if(isSuccessful(response)) {
+            //add api key to usage plan
+            CreateUsagePlanKeyRequest usagePlanKeyRequest = new CreateUsagePlanKeyRequest()
+            usagePlanKeyRequest.keyId = response.id
+            usagePlanKeyRequest.keyType = "API_KEY"
+            usagePlanKeyRequest.usagePlanId = usagePlanId
+            apiGatewayIdp.createUsagePlanKey(usagePlanKeyRequest)
+
+            return [apikeys:getApikeys(currentUser.userId), error: null]
+        }
+        else{
+            return [apikeys:null, error: "Could not generate api key"]
+        }
+    }
+
+    @Override
+    def getApikeys(String userId) {
+
+        GetApiKeysRequest getApiKeysRequest = new GetApiKeysRequest().withCustomerId(userId).withIncludeValues(true)
+        GetApiKeysResult response = apiGatewayIdp.getApiKeys(getApiKeysRequest)
+        if(isSuccessful(response)){
+            return response.items.value
+        }
+        else{
+            return null
+        }
+    }
+
+    @Override
+    def generateClient(String userId, List<String> callbackURLs, boolean forGalah){
+        CreateUserPoolClientRequest request =  new CreateUserPoolClientRequest().withUserPoolId(poolId)
+        request.clientName = "Client for user " + userId
```

Comment:
I'd like this link to be more formal and the client id should be less guessable... take a look at the CAS management code for how it auto generates client ids for ideas.  An AWS simpledb table to map userid to clientid should do. This could also be extracted to a separate service, IMHO.

Reply:
sure, btw this is the name of the client. not the client id. Client id is auto generated by the cognito